File: TextAlgorithms/K0412/Task3/Task3.py
def ReadTextData(file):
	'''
	N
	Pattern1
	.
	.
	.
	PatternN
	'''
	with open(file, 'r') as f:
		N = int(f.readline().strip())
		patterns = []
		for i in xrange(0, N):
			patterns.append(f.readline().strip())

	return N, patterns

def PrefixFunction(pattern):
	border = 0
	s = [0]*(len(pattern))
	for i in xrange(1, len(pattern)):
		while (border > 0) and (pattern[i] != pattern[border]):
			border = s[border - 1]
		if pattern[i] == pattern[border]:
			border += 1
		else:
			border = 0
		s[i] = border
	return s

def kmp(P, T):
	String = P + '$' + T
	s = PrefixFunction(String)
	result = list()
	for i in xrange(len(P) + 1, len(String)):
		if s[i] == len(P):
			result.append(i - 2 * len(P))
			break
	return result


N, patterns = ReadTextData('littera.in')
#patterns = ['abcd', 'abc','bce', 'bcd'] # ['abc','bcd','bce', 'abcd']
ans = 'NO'
patterns.sort(key=lambda x: (len(x), x.lower()))
#print(patterns)
for i in range(0, len(patterns)):
	pattern = patterns[0]
	patterns.pop(0)
	text = '$'.join(patterns)
	if kmp(pattern, text):
		ans = 'YES'
		break

with open('littera.out','w+') as f:
	f.write(str(ans))

File: TextAlgorithms/K0412/Task3/Task3.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>	
#include <algorithm>
#include <sstream>
using namespace std;

struct compare {
    bool operator()(const string& first, const string& second) {
        return first.size() < second.size();
    }
};


vector<int> PrefixFunction(string pattern){
	int N = (int) pattern.length();
	int border = 0;
	vector<int> s (N);
	for (int i = 1; i < N; i++){
		while (border > 0 && (pattern[i] != pattern[border])){
			border = s[border - 1];
		}
		if (pattern[i] == pattern[border]){
			border += 1;
		}
		else{
			border = 0;
		}
		s[i] = border;
	};
	return s;
};


vector<int>  kmp(string P, string T){
	int Plen = (int) P.length();
	string Str = P + '$' + T;
	int Strlen = (int) Str.length();
	vector<int> s = PrefixFunction(Str);
	vector<int> result;
	for (int i = Plen + 1; i < Strlen; i++){
		if (s[i] == Plen) {
			result.push_back(i - 2 * Plen);
			break;
		}
	};
	return result;
};

int main () {
int N;
compare c;
string ans = "NO"; 
ifstream input ("input.txt");
ofstream output ("output.txt");
input >> N;
vector<string> patterns (N);
for (int i = 0; i < N; i++) {
	 input >> patterns[i];
}
// cout << N << '\n';
// for (int i = 0; i < N; i++) {
// 	cout << patterns[i] << '\n';
// }
// cout << "Sorted" << '\n';
sort(patterns.begin(), patterns.end(), c);
string pattern;
string text;
ostringstream ss;
vector<int> result;
for (int i = 0; i < N; i++){
	pattern = patterns[0];
	patterns.erase(patterns.begin());
	for(size_t j = 0; j < patterns.size(); ++j){
  		if(j != 0)
    		ss << "$";
  		ss << patterns[j];
	}
	text = ss.str();
	result = kmp(pattern, text);
	if (!result.empty()){
		ans = "YES";
		break;
	};
		

};
output << ans;
}